src/charm.py

```diff
@@ -5,44 +5,60 @@
+    def __init__(self, framework):
+        super().__init__(framework)
         self.pebble_service_name = "fastapi-service"
-        self.container = self.unit.get_container("demo-server")  # see 'containers' in metadata.yaml
+        self.container = self.unit.get_container(
+            "demo-server"
+        )  # see 'containers' in charmcraft.yaml
 
         # Charm events defined in the database requires charm library.
         self.database = DatabaseRequires(self, relation_name="database", database_name="names_db")
-        self.framework.observe(self.database.on.database_created, self._on_database_created)
-        self.framework.observe(self.database.on.endpoints_changed, self._on_database_created)
-        self.framework.observe(self.on.database_relation_broken, self._on_database_relation_removed)
-
-        self.framework.observe(self.on.demo_server_pebble_ready, self._update_layer_and_restart)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.database.on.database_created, self._on_database_created)
+        framework.observe(self.database.on.endpoints_changed, self._on_database_created)
+
+        framework.observe(self.on.demo_server_pebble_ready, self._update_layer_and_restart)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+
+    def _on_collect_status(self, event):
+        port = self.config["server-port"]
+        if port == 22:
+            event.add_status(ops.BlockedStatus("Invalid port number, 22 is reserved for SSH"))
+        if not self.model.get_relation("database"):
+            # We need the user to do 'juju integrate'.
+            event.add_status(ops.BlockedStatus("Waiting for database relation"))
+        elif not self.database.fetch_relation_data():
+            # We need the charms to finish integrating.
+            event.add_status(ops.WaitingStatus("Waiting for database relation"))
+        try:
+            status = self.container.get_service(self.pebble_service_name)
+        except (ops.pebble.APIError, ops.ModelError):
+            event.add_status(ops.MaintenanceStatus("Waiting for Pebble in workload container"))
+        else:
+            if not status.is_running():
+                event.add_status(ops.MaintenanceStatus("Waiting for the service to start up"))
+        # If nothing is wrong, then the status is active.
+        event.add_status(ops.ActiveStatus())
 
     def _on_config_changed(self, event):
-        port = self.config["server-port"]  # see config.yaml
+        port = self.config["server-port"]  # see charmcraft.yaml
         logger.debug("New application port is requested: %s", port)
 
-        if int(port) == 22:
-            self.unit.status = BlockedStatus("Invalid port number, 22 is reserved for SSH")
+        if port == 22:
+            logger.info("Invalid port number, 22 is reserved for SSH")
```

Comment:
Maybe add a comment, something like `# BlockedStatus set in collect-status handler`.